### PR TITLE
Construct Synthetic Contract Result

### PR DESCRIPTION
**Pull request by georgi-l95**
**Description**:
To be able for the graph to access and ingest the events, they need first to be externalised in the contract result table. But as we only have them in the contract logs table, we have to find a way to synthetically construct them for this tool.

1. The Graph asks for every block by `getBlockByHash`, this is were we need to get all the synthetic logs as well. But as they might be duplicates, we need to filter them out and add them (depending if showDetails is true or not) to the transactionObject or transactionHashes array. When adding them to the transactionObjects array, we need to add them as transactions, but as we don't have all the necessary information to build a proper transaction object, we need to fake some of it (hence the magic numbers). All faked information here is random, The Graph is not using it (it's only need the transaction hash, which is correct), but it needs it to be in proper format (not undefined or odd length). We cache all synthetic logs as well to be able to use them later in getTransactionReceipt.

2. After it has all the transactions, it then ask for each one by getTransactionReceipt. That's why now we need to check if there are any cached synthetic logs by hash, apart from the other checks in this method. If there are get them and build a receipt.

If you want to test it out, start the relay using this branch and switch to branch `1040-expand-subgraph-example-with-hapi-events` and run the subgraph according to the readme in the tools/subgraph-example folder. Should be able to see in the logs as well as in the graphql playground that events are properly ingested.

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: packages/relay/src/lib/eth.ts [hunk 1]
```diff
@@ -1443,12 +1444,39 @@ export class EthImpl implements Eth {
...
+        logs: [cachedLog],
+        logsBloom: EthImpl.emptyBloom,
+        transactionHash: cachedLog.transactionHash,
+        transactionIndex: cachedLog.transactionIndex,
+        effectiveGasPrice: null,
+        root: null,
+        status: null,
+      };
+      this.logger.debug(`${requestIdPrefix} getTransactionReceipt returned cached synthetic receipt response: ${cachedResponse}`);
+      this.cache.set(cacheKey, receipt, EthImpl.ethGetTransactionReceipt, undefined, requestIdPrefix);
```
Comment (Nana-EC) [inline packages/relay/src/lib/eth.ts:1475]:
```suggestion
      this.cache.set(cacheKey, receipt, cacheKeySyntheticLog, syntheticLogCacheTtl, requestIdPrefix);
```

Would need to add the following to the top and set default to be like 1hr

```
private readonly syntheticLogCacheTtl =  parseNumericEnvVar('SYNTHETIC_LOG_CACHE_TTL', 'DEFAULT_SYNTHETIC_LOG_CACHE_TTL'));
```

File: packages/relay/src/lib/eth.ts [hunk 2]
```diff
@@ -1668,6 +1698,67 @@ export class EthImpl implements Eth {
...
+    return new Transaction({
+      accessList: undefined, // we don't support access lists for now, so punt
+      blockHash: log.blockHash,
+      blockNumber: log.blockNumber,
+      chainId: this.chain,
+      from: log.address,
+      gas: EthImpl.defaultTxGas,
+      gasPrice: undefined,
+      hash: log.transactionHash,
+      input: "0xabcd",
```
Comment (Nana-EC) [inline packages/relay/src/lib/eth.ts:1748]: Seems like `input` and `value` can be empty or just `0x`

**Review by Nana-EC — COMMENTED** [review]
Nice, some thoughts

File: packages/relay/src/lib/eth.ts [hunk 3]
```diff
@@ -1618,14 +1646,15 @@ export class EthImpl implements Eth {
     const blockResponse = await this.getHistoricalBlockResponse(blockHashOrNumber, true, requestIdPrefix);
 
     if (blockResponse == null) return null;
-
     const timestampRange = blockResponse.timestamp;
     const timestampRangeParams = [`gte:${timestampRange.from}`, `lte:${timestampRange.to}`];
     const contractResults = await this.mirrorNodeClient.getContractResults({ timestamp: timestampRangeParams }, undefined, requestIdPrefix);
     const maxGasLimit = constants.BLOCK_GAS_LIMIT;
     const gasUsed = blockResponse.gas_used;
+    const blockNumber = blockResponse.number.toString(16);
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply packages/relay/src/lib/eth.ts]: Removed extra block call so this isn't needed

**Review by AlfredoG87 — DISMISSED** [review]
LGTM

**Comment by AlfredoG87** [comment]
@georgi-l95 I tried to run both the relay and the graph-node example.

I noticed that we now are deploying HTS based ERC20 and  ERC721 Tokens, however, when starting the graph node am getting the next error.

I don't think is related to synthetic transaction logs, since I tried against `testnet` directly on hashIO and had the same issue.

```
subgraph-example-graph-node-1  | Jul 19 21:49:14.910 WARN Trying again after eth_getBlockByHash RPC call for block hash 0xbb0daf6d6637822da05f031b03a158916d0c69ac44bf871e458c7ce6d40680a8 failed (attempt #10) with result Err(Decoder error: Error("Invalid hex: Odd number of digits", line: 0, column: 0)), provider: local-rpc-0, component: BlockIngestor
```

However my local graph-node was not able to index anything due to that and I was not able to validate that the `syntethic transaction logs` are being indexed correctly.

Any ideas on what am I missing or what is the cause of the error pasted above?

**Comment by georgi-l95** [comment]
@AlfredoG87 Hi, yes, somewhere here the graph is not liking the value https://github.com/hashgraph/hedera-json-rpc-relay/pull/1518/files#diff-dd8c7990f7764960d05dcb4c1b5ae7eee48b425b1b83d6dcc7e7ccc442330ef0R1750-R1771
Try adjusting them, or see where they are odd lenght. I've adjusted them in my last commit and it was working, you can check it.

File: packages/relay/src/lib/eth.ts [hunk 4]
```diff
@@ -1443,12 +1444,39 @@ export class EthImpl implements Eth {
...
+        transactionIndex: cachedLog.transactionIndex,
+        effectiveGasPrice: null,
+        root: null,
+        status: null,
+      };
+      this.logger.debug(`${requestIdPrefix} getTransactionReceipt returned cached synthetic receipt response: ${cachedResponse}`);
+      this.cache.set(cacheKey, receipt, EthImpl.ethGetTransactionReceipt, undefined, requestIdPrefix);
+
+      return receipt
+    }
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply packages/relay/src/lib/eth.ts:1485]: If they are not cached beforehand in the getBlockByHash, they won't be available. I was thinking how we can retrieve them without caching, but wihout some API changes I'm not sure is possible.

[the remark this answers is not in the document]
Comment (Nana-EC) [reply packages/relay/src/lib/eth.ts:1485]: Yeah. Still thinking on this myself. 🤔 
Maybe a limitation for a future PR

**Comment by Nana-EC** [comment]
> @AlfredoG87 Hi, yes, somewhere here the graph is not liking the synthetic values that we pass https://github.com/hashgraph/hedera-json-rpc-relay/pull/1518/files#diff-dd8c7990f7764960d05dcb4c1b5ae7eee48b425b1b83d6dcc7e7ccc442330ef0R1758-R1770 Try adjusting them, or see where they are odd lenght. I've adjusted them in my last commit and it was working, you can check it.

Okay. This might have been me. I reverted some values cause they were arbitrary and i was hoping emptyHex was sufficient. May zeroHex in some cases.

We may need some more testing on the sub-graph side to be able to catch things like this.
I guess your PR #1040 might address that

**Comment by AlfredoG87** [comment]
@Nana-EC 

I was able to successfully ingest and index HTS Synthetic logs on grap-node and then query them.
image

**Review by AlfredoG87 — APPROVED** [review]
Is working and indexing correctly based on my tests
LGTM

**Pull request merged**